### PR TITLE
Adds support for non AWS S3 backend

### PR DESCRIPTION
**Pull request by stephanh**
This adds the ability to configure another Item Storage backend that uses S3 to store the cache in non AWS S3 backends such as Ceph or Swift. It has been tested with an S3 frontend to Swift.

Browsing the cache is currently not supported. I'm not sure how to enable that in a generic way at this stage.

Here is a screenshot of the configuration:
screen shot 2017-08-22 at 10 48 40

### CODE REVIEW

**Comment by stephanh** [comment]
@seanradley they were debug messages I forgot about. I have removed them.

File: src/main/resources/jenkins/plugins/itemstorage/s3/NonAWSS3ItemStorage/help-credentialsId.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  - IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  - FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  - AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  - LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  - OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  - THE SOFTWARE.
+  -->
+
+<div>
+    Select the AWS credentials that should be used to authenticate.  If none specified, will attempt to use instance
```
Comment (jsok) [inline src/main/resources/jenkins/plugins/itemstorage/s3/NonAWSS3ItemStorage/help-credentialsId.html]: For non-AWS this doesn't make sense -- if you don't set credentials there is no instance metadata or role to resort to. So in this mode the credentials should be required.

File: src/main/java/jenkins/plugins/itemstorage/s3/ClientHelper.java [hunk 2]
```diff
@@ -46,19 +59,30 @@ public ClientHelper(AWSCredentials credentials, String region, ProxyConfiguratio
...
+            System.out.println("Path Style Access: " + pathStyleAccess);
+            System.out.println("SignerVersion: " + signerVersion);
 
-            if (getCredentials() == null) {
-                client = new AmazonS3Client(getClientConfiguration(proxy));
-            } else {
-                client = new AmazonS3Client(getCredentials(), getClientConfiguration(proxy));
+            if (getCredentials() != null) {
+                builder.setCredentials(new AWSStaticCredentialsProvider(getCredentials()));
+                            }
```
Comment (stephanh) [inline src/main/java/jenkins/plugins/itemstorage/s3/ClientHelper.java]: Fixed.

File: src/main/resources/jenkins/plugins/itemstorage/s3/NonAWSS3ItemStorage/help-credentialsId.html (hunk 1 above)
Comment (stephanh) [inline src/main/resources/jenkins/plugins/itemstorage/s3/NonAWSS3ItemStorage/help-credentialsId.html]: Changed to require credentials and the help has been updated.

**Comment by stephanh** [comment]
@jsok I have changed the help message and drop down menu name.

**Pull request merged**